### PR TITLE
Memory swap

### PR DESCRIPTION
**Pull request by EricPyZhou**
`ptrSizeStore` is a vector of a structure named `DataElement` that contain device memory pointer, host memory pointer and size.
 
`swap_out`:
allocate host memory, copy everything element in `ptrSizeStore` into host, **free device memory**.

`swap_in`:
allocate device memory in the original location, copy everything from host to device, **free host memory**.

### CODE REVIEW

**Comment by EricPyZhou** [comment]
I will remove `test_swap.cu` and `test_swap` once review is done.

and Error handling is needed in the future to handle, for example, when `swap_in` is called before `swap_out` and there are elements in `ptrSizeStore` that contain nullptr of `hostPtr` because of executing cudaMalloc(): https://github.com/spcl/mignificient/blob/1c2046d5605449cac649ab68e226648cc2fdffa7/gpuless/trace/manager/manager_device.cpp#L85

File: gpuless/trace/manager/manager_device.cpp [hunk 1]
```diff
@@ -148,6 +182,38 @@ flatbuffers::FlatBufferBuilder handle_execute_request(const gpuless::FBProtocolM
...
+    ptrSizeStore.clear();
+}
+
+void swap_out() {
+    for (auto& buffer : ptrSizeStore) {
+        void* tempVector;
+        cudaHostAlloc(&tempVector, buffer.size, cudaHostAllocDefault);
+        buffer.hostPtr = tempVector;
+        // Schedule transfers from device to host
+        cudaMemcpy(buffer.hostPtr, buffer.devicePtr, buffer.size, cudaMemcpyDeviceToHost);
```
Comment (mcopik) [inline gpuless/trace/manager/manager_device.cpp]: Please use here async copy. Default memcpy will synchronize after every copy.

File: gpuless/trace/manager/manager_device.cpp [hunk 2]
```diff
@@ -148,6 +182,38 @@ flatbuffers::FlatBufferBuilder handle_execute_request(const gpuless::FBProtocolM
     return builder;
 }
 
+void swap_in() {
+    // Reserve memory on device
+    for (auto& buffer : ptrSizeStore) {
+        // allocate size
+        cudaMalloc(&buffer.devicePtr, buffer.size);
+        cudaMemcpy(buffer.devicePtr, buffer.hostPtr, buffer.size, cudaMemcpyHostToDevice);
```
Comment (mcopik) [inline gpuless/trace/manager/manager_device.cpp]: as below - async copies :)

**Review by mcopik — CHANGES_REQUESTED** [review]

**Comment by mcopik** [comment]
Thanks! I had to manually apply your changes since gpuless has been moved to a submodule. I used your implementation and extended it with full memory virtualization.

https://github.com/spcl/gpuless/commit/8fc7af0a8f8d3d795190df1a1848355952402475

**Pull request closed without merging**